### PR TITLE
feat(core): upload collection, truncate long file name with ellipsis in middle

### LINKED ISSUE #7623: File Uploader enhancement for truncating file names in the middle
#### Is this a bug, enhancement, or feature request?
Enhancement to an existing component
#### Briefly describe your proposal.
Requested by Erico:
Screen Shot 2022-01-11 at 10 47 58 AM

### PR DESCRIPTION
**Pull request by StevenZhao318**
## Related Issue(s)

closes #7623

## Description

Adds middle ellipsis to truncate file names that are too long. 

### Before:
image

### After:
image

##### During Implementation

1. Visual Testing:

-   [x] visual misalignments/updates
-   [x] check Light/Dark/HCB/HCW themes
-   [x] RTL/LTR - proper rendering and labeling
-   [x] responsiveness(resize)
-   [x] Content Density (Cozy/Compact/(Condensed))
-   [n/a] States - hover/disabled/focused/active/on click/selected/selected hover/press state
-   [x] Interaction/Animation - open/close, expand/collapse, add/remove, check/uncheck
-   [n/a] Mouse vs. Keyboard support
-   [x] Text Truncation

2. API and functional correctness

-   [x] check for console logs (warnings, errors)
-   [n/a] API boundary values
-   [n/a] different combinations of components - free style
-   [n/a] change the API values during testing

3. Documentation and Example validations

-   [n/a] missing API documentation or it is not understandable
-   [x] poor examples
-   [n/a] Stackblitz works for all examples

4. Accessibility testing
5. Browser Testing - Edge, Safari, Chrome, Firefox

##### PR Quality

-   [x] the commit message(s) follows the guideline:
        https://github.com/SAP/fundamental-ngx/blob/main/CONTRIBUTING.md
-   [n/a] tests for the changes that have been done
-   [x] all items on the PR Review Checklist are addressed :
        https://github.com/SAP/fundamental-ngx/wiki/PR-Review-Checklist
-   [n/a] Run npm run build-pack-library and test in external application
-   [n/a] update `README.md`
-   [n/a] [Breaking Changes Wiki](https://github.com/SAP/fundamental-ngx/wiki/Breaking-Changes)

### CODE REVIEW

File: libs/core/src/lib/upload-collection/upload-collection-item.directive.ts [hunk 1]
```diff
@@ -48,28 +60,76 @@ export class UploadCollectionItemDirective implements AfterContentInit, OnDestro
...
+            }
+        } else if (this.previousContainerWidth && this.containerWidth > this.previousContainerWidth) {
+            // Looks to expand file title if container width is resized to greater than before
+            // reset file title to initial
+            this._titleDirective.elRef.nativeElement.innerHTML = this.fileNameFull;
+            this.resizeFileTitle(this.fileNameFull);
+        }
+
+        this.previousContainerWidth = this.containerWidth;
+        this._cdRef.detectChanges();
```
Comment (mikerodonnell89) [inline libs/core/src/lib/upload-collection/upload-collection-item.directive.ts:120]: no need to detectChanges here and also we can remove the DI in the constructor

File: libs/core/src/lib/upload-collection/upload-collection-item.directive.ts [hunk 2]
```diff
@@ -48,28 +60,76 @@ export class UploadCollectionItemDirective implements AfterContentInit, OnDestro
...
+        if (!this.elementRef.nativeElement.parentElement) {
+            return;
+        }
+        this.titleWidth = this.getTitleWidth();
+        this.containerWidth = this.getContainerWidth();
+
+        // if first load and no previous container width, or if container boundary is resized to smaller than before
+        if (!this.previousContainerWidth || this.containerWidth < this.previousContainerWidth) {
+            // and the title extends past the container
+            if (this.titleWidth * 1.05 >= this.containerWidth) {
```
Comment (mikerodonnell89) [inline libs/core/src/lib/upload-collection/upload-collection-item.directive.ts:109]: Why not just `if (this.titleWidth >= this.containerWidth)`

Comment (StevenZhao318) [reply libs/core/src/lib/upload-collection/upload-collection-item.directive.ts:109]: explained above, i find that the button-group shifts when the text is nearly identical to the container size, so added some wiggle room.

File: libs/core/src/lib/upload-collection/upload-collection-item.directive.ts [hunk 3]
```diff
@@ -120,4 +180,50 @@ export class UploadCollectionItemDirective implements AfterContentInit, OnDestro
...
+     *
+     * truncates the string by cutting out excess length in the middle and replacing with '...'
+     */
+    private truncateTitle(str: string): string {
+        const cutLength = Math.floor(str.length * 0.8);
+
+        if (str.length > cutLength) {
+            const stringLeftIndex = Math.floor(cutLength / 2);
+            const stringRightIndex = str.length - stringLeftIndex + 3;
+            return str.substring(0, stringLeftIndex) + '... ' + str.substring(stringRightIndex, str.length);
```
[the remark this answers is not in the document]
Comment (mikerodonnell89) [reply libs/core/src/lib/upload-collection/upload-collection-item.directive.ts:203]: Looks a bit wonky to me but no strong opinion here either. for what it's worth, my mac isn't adding spaces when truncating in finder between words

**Review by mikerodonnell89 — APPROVED** [review]

**Pull request merged**